# TensorFlow

### Данные урока

In [111]:
from __future__ import absolute_import, division, print_function, unicode_literals

# TensorFlow and tf.keras
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras.layers import TextVectorization

# Helper libraries
import numpy as np
import matplotlib.pyplot as plt

print(tf.__version__)

2.11.0


## Практическое задание

<ol>
    <li>1. Попробуйте улучшить работу нейронной сети(разобранную на уроке) обучавшейся на датасет Fashion-MNIST. 
       Опишите в комментарии к уроку - какого результата вы добились от нейросети? Что помогло вам улучшить ее точность?<br><br>
    </li>
    
    <li>2. Поработайте с документацией TensorFlow 2. Попробуйте найти полезные команды TensorFlow неразобранные на уроке.</li>
        
    <li>*3. Попробуйте обучить нейронную сеть на TensorFlow 2 на датасете imdb_reviews. Опишите в комментарии к уроку - какой результата вы добились от нейросети? Что помогло вам улучшить ее точность? </li>

    
    
</ol>

In [2]:
np.random.seed(0)

In [3]:
from keras.datasets import imdb

In [93]:
imdb = keras.datasets.imdb # загрузка IMDB

(train_data, train_labels), (test_data, test_labels) = imdb.load_data(num_words=5000)

In [94]:
print("Training entries: {}, labels: {}".format(len(train_data),
                                               len(train_labels)))

Training entries: 25000, labels: 25000


In [95]:
print(train_data[0])

[1, 14, 22, 16, 43, 530, 973, 1622, 1385, 65, 458, 4468, 66, 3941, 4, 173, 36, 256, 5, 25, 100, 43, 838, 112, 50, 670, 2, 9, 35, 480, 284, 5, 150, 4, 172, 112, 167, 2, 336, 385, 39, 4, 172, 4536, 1111, 17, 546, 38, 13, 447, 4, 192, 50, 16, 6, 147, 2025, 19, 14, 22, 4, 1920, 4613, 469, 4, 22, 71, 87, 12, 16, 43, 530, 38, 76, 15, 13, 1247, 4, 22, 17, 515, 17, 12, 16, 626, 18, 2, 5, 62, 386, 12, 8, 316, 8, 106, 5, 4, 2223, 2, 16, 480, 66, 3785, 33, 4, 130, 12, 16, 38, 619, 5, 25, 124, 51, 36, 135, 48, 25, 1415, 33, 6, 22, 12, 215, 28, 77, 52, 5, 14, 407, 16, 82, 2, 8, 4, 107, 117, 2, 15, 256, 4, 2, 7, 3766, 5, 723, 36, 71, 43, 530, 476, 26, 400, 317, 46, 7, 4, 2, 1029, 13, 104, 88, 4, 381, 15, 297, 98, 32, 2071, 56, 26, 141, 6, 194, 2, 18, 4, 226, 22, 21, 134, 476, 26, 480, 5, 144, 30, 2, 18, 51, 36, 28, 224, 92, 25, 104, 4, 226, 65, 16, 38, 1334, 88, 12, 16, 283, 5, 16, 4472, 113, 103, 32, 15, 16, 2, 19, 178, 32]


In [96]:
len(train_data[2]), len(train_data[3]) # кол-во слов в третьем и четвертом отзывах

(141, 550)

In [97]:
len(train_data[0]), len(train_data[1]) 

(218, 189)

In [98]:
word_index = imdb.get_word_index() 

In [99]:
list(word_index.items())[:15]

[('fawn', 34701),
 ('tsukino', 52006),
 ('nunnery', 52007),
 ('sonja', 16816),
 ('vani', 63951),
 ('woods', 1408),
 ('spiders', 16115),
 ('hanging', 2345),
 ('woody', 2289),
 ('trawling', 52008),
 ("hold's", 52009),
 ('comically', 11307),
 ('localized', 40830),
 ('disobeying', 30568),
 ("'royale", 52010)]

In [102]:
word_index = {k:(v+3) for k,v in word_index.items()}
word_index[""] = 0
word_index[""] = 1
word_index[""] = 2 
word_index[""] = 3

reverse_word_index = dict([(value, key) for (key, value) in word_index.items()])

def decode_review(text):
    return ' '.join([reverse_word_index.get(i, '?') for i in text])

In [103]:
decode_review(train_data[0]) # преобразование чисел в слова

"? this film was just brilliant casting location scenery story direction everyone's really suited the part they played and you could just imagine being there robert ? is an amazing actor and now the same being director ? father came from the same scottish island as myself so i loved the fact there was a real connection with this film the witty remarks throughout the film were great it was just brilliant so much that i bought the film as soon as it was released for ? and would recommend it to everyone to watch and the fly ? was amazing really cried at the end it was so sad and you know what they say if you cry at a film it must have been good and this definitely was also ? to the two little ? that played the ? of norman and paul they were just brilliant children are often left out of the ? list i think because the stars that play them all grown up are such a big ? for the whole film but these children are amazing and should be ? for what they have done don't you think the whole story wa

In [104]:
maxlen = 120

train_data = keras.preprocessing.sequence.pad_sequences(train_data, maxlen=maxlen)
test_data = keras.preprocessing.sequence.pad_sequences(train_data, maxlen=maxlen)

In [106]:
len(train_data[0]), len(train_data[1])

(120, 120)

In [108]:
model = keras.Sequential()
model.add(keras.layers.Embedding(5000, 16))
model.add(keras.layers.GlobalAveragePooling1D())
model.add(keras.layers.Dense(16, activation=tf.nn.relu))
model.add(keras.layers.Dense(1, activation=tf.nn.sigmoid))

model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding (Embedding)       (None, None, 16)          80000     
                                                                 
 global_average_pooling1d (G  (None, 16)               0         
 lobalAveragePooling1D)                                          
                                                                 
 dense_1 (Dense)             (None, 16)                272       
                                                                 
 dense_2 (Dense)             (None, 1)                 17        
                                                                 
Total params: 80,289
Trainable params: 80,289
Non-trainable params: 0
_________________________________________________________________


In [107]:
inputs = keras.Input(shape=(None, None, 3))
processed = keras.layers.RandomCrop(width=32, height=32)(inputs)
conv = keras.layers.Conv2D(filters=2, kernel_size=3)(processed)
pooling = keras.layers.GlobalAveragePooling2D()(conv)
feature = keras.layers.Dense(10)(pooling)

full_model = keras.Model(inputs, feature)
backbone = keras.Model(processed, conv)
activations = keras.Model(conv, feature)

In [116]:
model.compile(optimizer='adam',
              loss='binary_crossentropy',
              metrics=['accuracy'])

In [122]:
# обучение нейросети

model.fit(train_data,
          train_labels,
          epochs=30,
          batch_size=300,
          verbose=1)

Epoch 1/30
84/84 [==============================] - 1s 9ms/step - loss: 0.1508 - accuracy: 0.9475
Epoch 2/30
84/84 [==============================] - 1s 10ms/step - loss: 0.1498 - accuracy: 0.9488
Epoch 3/30
84/84 [==============================] - 1s 10ms/step - loss: 0.1497 - accuracy: 0.9486
Epoch 4/30
84/84 [==============================] - 1s 10ms/step - loss: 0.1495 - accuracy: 0.9492
Epoch 5/30
84/84 [==============================] - 1s 11ms/step - loss: 0.1493 - accuracy: 0.9488
Epoch 6/30
84/84 [==============================] - 1s 10ms/step - loss: 0.1492 - accuracy: 0.9492
Epoch 7/30
84/84 [==============================] - 1s 10ms/step - loss: 0.1484 - accuracy: 0.9499
Epoch 8/30
84/84 [==============================] - 1s 10ms/step - loss: 0.1480 - accuracy: 0.9491
Epoch 9/30
84/84 [==============================] - 1s 10ms/step - loss: 0.1473 - accuracy: 0.9508
Epoch 10/30
84/84 [==============================] - 1s 10ms/step - loss: 0.1480 - accuracy: 0.9494
Epoch 11/3

In [123]:
# результаты на тестовом наборе

results = model.evaluate(test_data, test_labels)

print(results)

782/782 [==============================] - 2s 2ms/step - loss: 3.2166 - accuracy: 0.4963
[3.216641902923584, 0.4962800145149231]


#### ответы:

*1. точность нейросети зависит от количества epochs и batch_size, чем больше тем выше accuracy* 

*2. на тестовом наборе вышло accuracy = 0.496*